### PR TITLE
698: revise deduped hearings list

### LINKED ISSUE #698: Revise Deduping hearing list
- refactor to remove code duplication--create component
- include list of actions for each hearing in list
- different deduped lists for recommendation form vs. for list on show-project and to-review

### PR DESCRIPTION
**Pull request by trbmcginnis**
**New Component**
- put `deduped-hearings-list` into separate component
- the deduped hearings list now concatenates action model objects for duplicate hearings --> these actions can be accessed as a new property called `hearingActions` on the `dedupedHearings` array 
- hearings list component displayed on `show-project`, `to-review`, and confirmation screen for hearing form

**New Helper**
- `hearings-submitted` is a new helper that checks whether every disposition for a project has `publichearinglocation` and `dateofpublichearing`

**Updated Tests**
- tests filled out for `date-time-picker`, `deduped-hearings-list`, `hearings-submitted`, and `user-can-fill-out-hearing-form`

TODO 1: continue to revise these tests to capture more edge cases
TODO 2: better design the list of actions in the deduped hearings list -- captured in issue #712

Closes #698

### CODE REVIEW

File: app/components/deduped-hearings-list.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    // during the reduce, if there is a duplicate in the array of dispositions,
+    // the actions model for that duplicate disposition is pushed into this array
+    dispositions.forEach(function(disposition) {
+      disposition.set('hearingActions', [disposition.action]);
+    });
+
+    // hearingsSubmitted is set when the component is rendered
+    if (hearingsSubmitted) {
+      // reduce method to deduplicate dispositions based on location and date of hearing
+      deduped = dispositions.reduce((accumulatedDisps, currentDisp) => {
```
Comment (allthesignals) [inline app/components/deduped-hearings-list.js]: nit: this method is doing a lot. I like the idea of returning this reduce by itself from somewhere.

Comment (trbmcginnis) [reply app/components/deduped-hearings-list.js]: @allthesignals "returning this reduce by itself from somewhere" -- what do you mean by that? Separating out these functions?

File: app/templates/components/to-review-project-card.hbs [hunk 2]
```diff
@@ -22,33 +22,8 @@
...
-                      <span class="light-gray">|</span>
-                      <span class="hearing-time">
-                        {{~moment-format hearing.dateofpublichearing "h:mm A"~}}
-                      </span>
-                    </div>
-                  </li>
-                {{/each}}
-                </ul>
+          {{#if hearingsSubmittedForProject}}
+            {{deduped-hearings-list project=project hearingsSubmitted=hearingsSubmittedForProject attribute="ulurpnumber"}}
```
Comment (godfreyyeung) [inline app/templates/components/to-review-project-card.hbs]: sorry, another nit: We should put each parameter on its own line, e.g. 
```
   project=project 
   hearingsSubmitted=hearingsSubmittedForProject
   attribute="ulurpnumber"
}}
```

File: app/components/deduped-hearings-list.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        // object that represents a match between accumulatedDisps and the currentDisp
+        const matchingPropsObject = accumulatedDisps.find(item => item.publichearinglocation === currentDisp.publichearinglocation && item.dateofpublichearing.toString() === currentDisp.dateofpublichearing.toString());
+
+        // if an object exists in accumulatedDisps that matches the currentDisp
+        if (matchingPropsObject) {
+          // iterate through the array of accumulated dispositions
+          accumulatedDisps.forEach(function(disposition) {
+            // check whether an object in accumulatedDisps matches the currentDisp object
+            const matchingObject = disposition.publichearinglocation === currentDisp.publichearinglocation && disposition.dateofpublichearing.toString() === currentDisp.dateofpublichearing.toString();
```
Comment (godfreyyeung) [inline app/components/deduped-hearings-list.js]: Nice! Love the descriptive naming here. Could you just use the `matchingPropsObject` from line 33 here on like 38, instead of doing another find? 

It seems like `matchingPropsObject` is already a reference to the matching props object, so no need to get it again?

File: app/components/deduped-hearings-list.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+            if (matchingObject) {
+              // grab the actions model from the currentDisp
+              const currentDispAction = currentDisp.get('action');
+              // push this into the hearingActions array on the matched object in accumulatedDisps
+              disposition.hearingActions.push(currentDispAction);
```
Comment (godfreyyeung) [inline app/components/deduped-hearings-list.js]: awesome. Can we additionally associated each disp with its "duplicate disps"? I.e. add this line:
`disposition.duplicateDisps.push(curredDisp)`

 this will make it easier for the recommendation page to assign  a 'wasquorumpresent' value to each disposition binned under a hearing.

File: app/templates/components/deduped-hearings-list.hbs [hunk 5]
```diff
@@ -0,0 +1,35 @@
+<h5 class="column-header">Hearing Information</h5>
+    <ul class="no-bullet">
+    {{#each dedupedHearings as |hearing|}}
```
Comment (godfreyyeung) [inline app/templates/components/deduped-hearings-list.hbs]: awesome. Suggestion: To make this `dedupped-hearings-list` component re-usable and portable to the recommendation form, this component can just yield the `deduppedHearings` as a block parameter. 
For example 
```
{{yield (hash dedupedHearings=dedupedHearings)}}
```

And then all this markup below would be specific to the hearing form (hearing/add.hbs):
```
{{deduped-hearings-list as |dedupedHearings|}}
  <h5 class="column-header">Hearing Information</h5>
    <ul class="no-bullet">
   <li class="grid-x">
        <div class="cell shrink small-margin-right">
          {{#if (is-before hearing.dateofpublichearing)}}
            {{fa-icon "calendar" class="light-gray" fixedWidth=true}}
          {{else}}
            {{fa-icon "check" class="blue" fixedWidth=true}}
          {{/if}}
     ....
{{/deduped-hearings-list}}
```

This way, the recommendation form can add little radio boxes and such next to each dedupped hearing. 
This review comment may not be very clear, happy to chat in person.

Comment (godfreyyeung) [reply app/templates/components/deduped-hearings-list.hbs]: @trbmcginnis actually it would just be {{yield dedupedHearings}}, my bad

File: app/components/deduped-hearings-list.js (hunk 4 above)
Comment (trbmcginnis) [reply app/components/deduped-hearings-list.js]: will do

File: app/templates/components/deduped-hearings-list.hbs (hunk 5 above)
Comment (trbmcginnis) [reply app/templates/components/deduped-hearings-list.hbs]: will do this in another issue
